FLAnimatedImage/FLAnimatedImageFrameCache.m

```diff
@@ -0,0 +1,507 @@
+                }
+                FLLogVerbose(@"Predrew frame %lu in %f ms for animated image: %@", (unsigned long)i, (predrawDuration + slowdownDuration) * 1000, self);
+#endif
+                // The results get returned one by one as soon as they're ready (and not in batch).
+                // The benefits of having the first frames as quick as possible outweigh building up a buffer to cope with potential hiccups when the CPU suddenly gets busy.
+                if (image) {
+                    dispatch_async(dispatch_get_main_queue(), ^{
+                        [strongSelf cacheFrame:image atIndex:i];
+#if defined(DEBUG) && DEBUG
+                        if ([weakSelf.debug_delegate respondsToSelector:@selector(debug_animatedImage:didUpdateCachedFrames:)]) {
+                            [weakSelf.debug_delegate debug_animatedImage:nil didUpdateCachedFrames:strongSelf->_cachedFrameIndexes];
+                        }
+#endif
+                    });
+                }
+            }
+        };
+        
+        [frameIndexesToAddToCache enumerateRangesInRange:firstRange options:0 usingBlock:frameRangeBlock];
+        [frameIndexesToAddToCache enumerateRangesInRange:secondRange options:0 usingBlock:frameRangeBlock];
+    });
+}
+
+/**
+ * This must be called on the main thread.
+ */
+- (void)cacheFrame:(UIImage *)image atIndex:(NSUInteger)index
+{
+    NSAssert([NSThread isMainThread], @"This method must be called on the main thread");
+
+    NSMutableDictionary *mutableCachedFramesForIndexes = [_cachedFramesForIndexes mutableCopy];
+    mutableCachedFramesForIndexes[@(index)] = image;
+    [self setCachedFramesForIndexes:mutableCachedFramesForIndexes];
+    [_cachedFrameIndexes addIndex:index];
+    [_requestedFrameIndexes removeIndex:index];
+}
+
+// This method is called on _serialQueue. Renders the image at index `i`, taking
+// blending with previous frames into account.
+- (UIImage *)synchronousImageAtIndex:(NSUInteger)index
+{
+    // Lock _cachedFramesMutex since we are reading from it, and it could be concurrently being modified
+    // on the main thread.
+    pthread_mutex_lock(&_cachedFramesMutex);
+    UIImage *image = _cachedFramesForIndexes[@(index)];
+    pthread_mutex_unlock(&_cachedFramesMutex);
+
+    if (image) {
+        return image;
+    }
+
+
+
+    // Hopefully we won't hit the code path often, as it will be expensive if for frame N
+    // we need to render frames [0-(N-1)]. It is possible we'll hit this for particularly large
+    // images, or after a memory warning.
+    // TODO: Look into always caching (_requestedFrameIndex - 1) to reduce the likelihood of
+    // running into this code path.
+    image = [self.dataSource imageAtIndex:index];
+    if ([self.dataSource frameRequiresBlendingWithPreviousFrame:index]) {
+//        UIImage *previousImage = [self synchronousImageAtIndex:index - 1]; // NOTE: 加上外面的迴圈 這樣不就變成 n x n ??
+        UIImage *previousImage;
+        if (index != 0) {
```

Comment:
Should we also check for the upper bounds?

Reply:
Maybe it is not necessary. Because the index comes from this loop, we are not going hit outside the bounds.
https://github.com/cardinalblue/FLAnimatedImage/blob/19efbd546d7a3ccfb1afdae529bb9afe907f5dee/FLAnimatedImage/FLAnimatedImageFrameCache.m#L328